Detectors/TPC/workflow/src/RawToDigitsSpec.cxx

```diff
@@ -0,0 +1,209 @@
+
+using namespace o2::framework;
+using SubSpecificationType = o2::framework::DataAllocator::SubSpecificationType;
+
+namespace o2
+{
+namespace tpc
+{
+
+o2::framework::DataProcessorSpec getRawToDigitsSpec(int channel, const std::string_view inputDef)
+{
+
+  struct ProcessAttributes {
+    DigitDump digitDump;
+    rawreader::RawReaderCRUManager rawReader;
+    uint32_t lastOrbit{0};
+    uint32_t maxEvents{100};
+    bool quit{false};
+  };
+
+  auto initFunction = [channel](InitContext& ic) {
+    auto processAttributes = std::make_shared<ProcessAttributes>();
+    // set up calibration
+    {
+      auto& digitDump = processAttributes->digitDump;
+      digitDump.init();
+      digitDump.setInMemoryOnly();
+      const auto pedestalFile = ic.options().get<std::string>("pedestal-file");
+      LOG(INFO) << "Setting pedestal file: " << pedestalFile;
+      digitDump.setPedestalAndNoiseFile(pedestalFile);
+
+      processAttributes->rawReader.createReader("");
+      processAttributes->rawReader.setADCDataCallback([&digitDump](const PadROCPos& padROCPos, const CRU& cru, const gsl::span<const uint32_t> data) -> Int_t {
+        Int_t timeBins = digitDump.update(padROCPos, cru, data);
+        digitDump.setNumberOfProcessedTimeBins(std::max(digitDump.getNumberOfProcessedTimeBins(), size_t(timeBins)));
+        return timeBins;
+      });
+      processAttributes->maxEvents = static_cast<uint32_t>(ic.options().get<int>("max-events"));
+    }
+
+    auto processingFct = [processAttributes, channel](ProcessingContext& pc) {
+      if (processAttributes->quit) {
+        return;
+      }
+
+      uint64_t activeSectors = 0;
+
+      // lambda that snapshots digits to be sent out; prepares and attaches header with sector information
+      auto snapshotDigits = [&pc, activeSectors, channel](std::vector<o2::tpc::Digit> const& digits, int sector) {
+        o2::tpc::TPCSectorHeader header{sector};
+        header.activeSectors = activeSectors;
+        // note that snapshoting only works with non-const references (to be fixed?)
+        pc.outputs().snapshot(Output{"TPC", "DIGITS", static_cast<SubSpecificationType>(channel), Lifetime::Timeframe, header},
+                              const_cast<std::vector<o2::tpc::Digit>&>(digits));
+      };
+
+      // loop over all inputs
+      for (auto& input : pc.inputs()) {
+        const auto* dh = DataRefUtils::getHeader<o2::header::DataHeader*>(input);
+
+        // select only RAW data
+        if (dh->dataDescription != o2::header::gDataDescriptionRawData) {
+          continue;
+        }
+
+        const auto subSpecification = dh->subSpecification;
+        const auto cruID = subSpecification >> 16;
+        const auto linkID = ((subSpecification + (subSpecification >> 8)) & 0xFF) - 1;
+        const auto dataWrapperID = ((subSpecification >> 8) & 0xFF) > 0;
+        const auto globalLinkID = linkID + dataWrapperID * 12;
+        const auto sector = cruID / 10;
+        activeSectors |= 0x1 << sector;
```

Comment:
Hi Jens, you need to refine the logic for the `activeSector` variable. This is a bit field which should be calculated once to signal all active sectors to the tracker which will then wait for a complete data set. Right now it looks like you are setting only the bit for the current sector which is propagated anyhow as the sector member of the header. So for one TF entity you need to calculate the active sectors before sending the first digit output, and then the active sector member variable in the header needs to be the same for all output parts of the TF.

I think that's why the tracker does not run when connected directly to the raw workflow. But if you have only one sector it should work. That could be a quick test for this hypothesis.

Reply:
I have checked the logic in the tracker. It is taking the activeSector bitfield face value, meaning that within one computing it will start the processing if it has data for each of the corresponding sectors. In that respect it should work if you always only set one bit for the current sector.

I propose the following logic: the bitfield should always have ones for bits, during sending the active sectors should be accumulated and when the last block is sent, this bitfield must be transmitted. Having all bits set (also those which actually do not correspond to sectors) will make the tracker wait for more data. Only if the transmitted active sector bitfield matches the buffered data, the tracker will start.